### PR TITLE
[WIP] Updated API for MoleculeNet loader functions

### LINKED ISSUE #2192: Unify API for loader functions
The molnet loader functions are currently divided into two groups with inconsistent APIs for specifying featurizers and splitters.

Most of them have arguments `featurizer` and `split` to specify them.  These arguments take one of a set of hardcoded strings, like `featurizer='ECFP'` or `split='random'`.  There are a few problems here.

1. The list of accepted values is undocumented.
2. There's also no way to discover them programmatically.
3. There's no way to specify a splitter or featurizer that isn't on the list.
4. The list of allowed options varies widely between datasets.  If there's a coherent set of rules behind them, I don't know what it is.

Then there are the ones that use the template introduced in #1938.  These work rather differently.  The argument for specifying the splitter is called `splitter` instead of `split`.  The arguments may take either the name of a class, the class itself, or an instance of the class.  These functions have their own set of issues.

1. The list of accepted values is again undocumented.
2. It is possible to discover them programmatically (for example from `dc.molnet.load_function.zinc15_datasets.zinc15_splitters`), but the mechanism is itself undocumented.
3. Much of the documentation is incorrect.  For example,

https://github.com/deepchem/deepchem/blob/ea0fe592fe4228109ea955c97c9d6e06c280e5a1/deepchem/molnet/load_function/zinc15_datasets.py#L91-L92

In fact the value should specify a single splitter, not "allowed splitters".  And it never mentions the possibility of passing a string or class.

4. It isn't clear to me that there's really a lot of benefit from having so many options.  `featurizer='CircularFeaturizer'` isn't substantially clearer, shorter, or more convenient than `featurizer=dc.feat.CircularFeaturizer()`.

We should come up with a single consistent API for all of them.

### PR DESCRIPTION
**Pull request by peastman**
Fixes #2192.  I've updated only Delaney so far.  Please take a look and see if it looks ok.  Once we're all happy with the design, we can do all the other loader functions.

It took some work to get mypy to pass.  Some of that was because of it being inflexible, and some was genuine errors.  `train_valid_test_split()` allowed you to pass 0 for `frac_valid`.  In that case, it would only create two datasets and return None for the validation set.  This required all downstream code to deal with the possibility that one of the returned values could be None.  I couldn't find anywhere we actually use this feature, and if you don't want a validation set, that's what `train_test_split()` is for, so I removed it.  The second problem was that transforming a dataset isn't guaranteed to return a DiskDataset, even if that's what the original dataset is, but the loading code assumed it was.  I added checks for this.

As noted in https://github.com/deepchem/deepchem/issues/2192#issuecomment-707488680, the option `split=None` didn't entirely work correctly.  It took some restructuring of the code to fix it.  I also changed the return value in that case.  It now returns a single element tuple for the dataset instead of a three element tuple with two elements being None.

### CODE REVIEW

File: deepchem/molnet/load_function/delaney_datasets.py [hunk 1]
```diff
@@ -34,95 +38,102 @@ def load_delaney(featurizer='ECFP',
...
     data_dir = DEFAULT_DIR
   if save_dir is None:
     save_dir = DEFAULT_DIR
+  tasks = ['measured log solubility in mols per litre']
+
+  # Try to reload cached datasets.
+
   if reload:
-    save_folder = os.path.join(save_dir, "delaney-featurized")
+    featurizer_name = str(featurizer.__class__.__name__)
```
Comment (rbharath) [inline deepchem/molnet/load_function/delaney_datasets.py:83]: One thought here is perhaps we could stringify it more intelligently so we can store multiple versions that use the same featurizer but different parameters. Looking at the string representation:

```
In [5]: feat = dc.feat.CircularFingerprint(size=1024)                  

In [6]: feat                       
Out[6]: CircularFingerprint[radius=2, size=1024, chiral=False, bonds=True, features=False, sparse=False, smiles=False]
```
Perhaps this could be
```
featurizer_name = CircularFingerprint_radius_2_size_1024_chiral_False_bonds_True_features_False_sparse_False_smiles_False
```
And similarly for splitters

**Review by rbharath — COMMENTED** [review]
I really like the new style! It strikes a balance between the template version and the old version to find something clean. I've made a few suggestions below.

File: deepchem/molnet/defaults.py [hunk 2]
```diff
@@ -9,12 +9,29 @@
...
 from deepchem.trans.transformers import Transformer
 from deepchem.splits.splitters import Splitter
 
 logger = logging.getLogger(__name__)
 
+featurizers = {
+    'ECFP': dc.feat.CircularFingerprint(size=1024),
+    'GraphConv': dc.feat.ConvMolFeaturizer(),
+    'Weave': dc.feat.WeaveFeaturizer(),
+    'Raw': dc.feat.RawFeaturizer()
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/molnet/defaults.py]: Done.

File: deepchem/molnet/load_function/delaney_datasets.py (hunk 1 above)
Comment (peastman) [reply deepchem/molnet/load_function/delaney_datasets.py]: Done.  I copied over the implementations of `__str__()` and `__repr()__` from Featurizer to Splitter.

File: deepchem/splits/splitters.py [hunk 3]
```diff
@@ -274,7 +271,30 @@ def __str__(self) -> str:
     >>> str(dc.splits.RandomSplitter())
     'RandomSplitter'
     """
-    return self.__class__.__name__
+    args_spec = inspect.getfullargspec(self.__init__)  # type: ignore
```
Comment (rbharath) [inline deepchem/splits/splitters.py:274]: This is great! Will also make DeepChem nicer to use from the REPL :)

**Review by rbharath — APPROVED** [review]
Looks good to me! Feel free to merge whenever ready :)

**Comment by peastman** [comment]
> Looks good to me! Feel free to merge whenever ready :)

I've barely started!  Next step is to try to separate out all the generic code to minimize the amount that's specific to each dataset.  Then we need to update all the other loaders to work the same way.

**Comment by peastman** [comment]
Ready for the next review.  I refactored it to provide a generic loader class, so each dataset will need much less code.

**Comment by peastman** [comment]
Travis reports some tests failing, but it's not clear that they're related to this PR.  I'm seeing the same failures even on the master branch.

File: deepchem/molnet/load_function/delaney_datasets.py [hunk 4]
```diff
@@ -3,29 +3,52 @@
...
+  def create_dataset(self) -> Dataset:
+    logger.info("About to featurize Delaney dataset.")
+    dataset_file = os.path.join(self.data_dir, "delaney-processed.csv")
+    if not os.path.exists(dataset_file):
+      dc.utils.data_utils.download_url(url=DELANEY_URL, dest_dir=self.data_dir)
+    loader = dc.data.CSVLoader(
+        tasks=DELANEY_TASKS, feature_field="smiles", featurizer=self.featurizer)
+    return loader.create_dataset(dataset_file, shard_size=8192)
+
+  def get_transformers(self, dataset: Dataset) -> List[dc.trans.Transformer]:
```
Comment (rbharath) [inline deepchem/molnet/load_function/delaney_datasets.py:28]: To allow users to specify their own transformers, perhaps we can make `get_transformers` be the default if no user-specified transformations are applied.

File: deepchem/molnet/load_function/molnet_loader.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+
+class _MolnetLoader(object):
+  """The class provides common functionality used by many molnet loader functions.
+  It is an abstract class.  Subclasses implement loading of particular datasets.
+  """
+
+  def __init__(self, featurizer: Union[dc.feat.Featurizer, str],
+               splitter: Union[dc.splits.Splitter, str, None],
+               data_dir: Optional[str], save_dir: Optional[str], **kwargs):
```
Comment (rbharath) [inline deepchem/molnet/load_function/molnet_loader.py:37]: Perhaps we can add a `transformers` kwarg here to allow for user specified transformations.

File: deepchem/molnet/load_function/delaney_datasets.py [hunk 6]
```diff
@@ -34,95 +57,38 @@ def load_delaney(featurizer='ECFP',
...
-    valid = transformer.transform(valid)
-    test = transformer.transform(test)
-
-  if reload:
-    deepchem.utils.data_utils.save_dataset_to_disk(save_folder, train, valid,
-                                                   test, transformers)
-  return delaney_tasks, (train, valid, test), transformers
+  loader = _DelaneyLoader(
+      featurizer, splitter, data_dir, save_dir, move_mean=move_mean, **kwargs)
+  featurizer_name = str(loader.featurizer)
```
Comment (rbharath) [inline deepchem/molnet/load_function/delaney_datasets.py:88]: Should we just make a helper function to get the save folder name? The "_mean_unmoved" makes it a little awkward for Delaney  but this feels like a common pattern

**Review by rbharath — COMMENTED** [review]
This is really slick! I like the idea of minimizing the amount of custom code that each loader has. That will make this code much much easier to maintain going forward.

I have a couple comments about small generalizations and cleanup but liking what I'm seeing :)

**Comment by nd-02110114** [comment]
This is really awesome refactoring! And, I just have one comment. 
Is it better to add the test for the molnet function...? I know Nathan wrote some tests.
https://github.com/deepchem/deepchem/tree/master/deepchem/molnet/tests

File: deepchem/molnet/load_function/molnet_loader.py (hunk 5 above)
Comment (peastman) [reply deepchem/molnet/load_function/molnet_loader.py:37]: We can't pass in transformers, because some transformers require the dataset as a constructor argument.  That means the transformers can't be created until after the dataset has already been loaded.

File: deepchem/molnet/load_function/delaney_datasets.py (hunk 6 above)
Comment (peastman) [reply deepchem/molnet/load_function/delaney_datasets.py:88]: We really should have tests for all molnet loader functions.  But running all of them will take several hours at least and involve downloading many GB of data.  Figuring out a good strategy for testing molnet is beyond the scope of this PR.

**Comment by peastman** [comment]
We really should have tests for all molnet loader functions. But running all of them will take several hours at least and involve downloading many GB of data. Figuring out a good strategy for testing molnet is beyond the scope of this PR.

File: deepchem/molnet/load_function/delaney_datasets.py (hunk 6 above)
Comment (peastman) [reply deepchem/molnet/load_function/delaney_datasets.py:88]: I tried to come up with a generic way for setting the folder name, but couldn't think of anything clean.  The code here to assemble the path is five lines, and three of those five lines are in some way specific to this dataset.

File: deepchem/molnet/load_function/molnet_loader.py (hunk 5 above)
Comment (rbharath) [reply deepchem/molnet/load_function/molnet_loader.py:37]: Good point ! As a simple possibility, perhaps we could allow for a `no_transformers` option so users can obtain the dataset without any transformations applied? They can then apply their own transformations as desired

**Comment by rbharath** [comment]
@ncfrey could comment better on tests here, but I think the idea he's followed is to add dummy data files for the datasets that follow the format of the full dataset raw file, but allow for quick testing of the loader function pointed at this dummy file

**Comment by ncfrey** [comment]
> @ncfrey could comment better on tests here, but I think the idea he's followed is to add dummy data files for the datasets that follow the format of the full dataset raw file, but allow for quick testing of the loader function pointed at this dummy file

This refactoring is great @peastman! If you take a look [here](https://github.com/deepchem/deepchem/blob/master/deepchem/molnet/load_function/tests/test_load_zinc15.py), there's a test for the ZINC15 loader. It is very minimal, all it does is take a small slice of the dataset, load it locally and featurize. Maybe this could be improved with at least a check that the link to the dataset on the AWS bucket is valid. Happy to discuss more outside of this PR.

**Comment by peastman** [comment]
Here is a half formed idea for a design.  Instead of passing a list of Transformers, you'll pass a function that takes a Dataset and returns a list of Transformers.  We'll provide functions for all the common cases.  `load_delaney()` would declare the argument as

```python
transformers: Callable[[Dataset], List[Transformer]] = normalization_transformer
```

If you wanted something different, you could specify `transformers=minmax_transformer` or `transformers=None` or `transformers=my_function`.

The problem is how to reconcile this with reloading cached datasets?  We label them on disk by the featurizer and splitter, not the transformers.  Looking for a cached dataset is the first thing we do.  We haven't yet created a dataset, so we can't yet call the function, so we don't know what transformers it will create.

Maybe we could define a class that you pass in rather than a function?  A `TransformerGenerator` would define both `get_transformers()` to build the transformers and `get_directory_name()` to define where it's stored on disk?

This would also give us a much cleaner way of handling the `move_mean` argument, which currently is a hack for one special case of customizing the transformers.  We could get rid of that argument, and instead you would specify `transformers=NormalizationTransformerGenerator(move_mean=False)`.

**Comment by peastman** [comment]
I tested on a second computer and confirmed that the tests in `test_graph_models.py` are failing on the master branch.  I don't know why they didn't show up on travis until now.  Anyone have an idea what caused them?

**Comment by rbharath** [comment]
> I tested on a second computer and confirmed that the tests in `test_graph_models.py` are failing on the master branch. I don't know why they didn't show up on travis until now. Anyone have an idea what caused them?

I just tested `test_graph_models` on my local machine and I'm getting the following errors:

```
FAILED test_graph_models.py::test_dag_model - assert 0.8681318681318682 >= 0.9
FAILED test_graph_models.py::test_dag_regression_model - assert 0.462634768824231 < 0.15
FAILED test_graph_models.py::test_dag_regression_uncertainty - assert 0.5927464085171481 < (0.7 * 0.8101929641020245)
FAILED test_graph_models.py::test_mpnn_regression_model - assert 0.1311314735151417 < 0.1
```

One possibility is that the version bump to TF 2.3 might have changed some of the numerics? None of these models are failing "strongly," in that they're clearly learning something but not hitting the previous target for fit. I think these are slow tests which don't get run by Travis. My bad for not doing a full slow run before the version bump!

Perhaps to avoid this, we could set up another Travis CI branch to run the slow tests as we're starting to do for the examples. That could help us catch these errors.

> Maybe we could define a class that you pass in rather than a function? A `TransformerGenerator` would define both `get_transformers()` to build the transformers and `get_directory_name()` to define where it's stored on disk?
> 
> This would also give us a much cleaner way of handling the `move_mean` argument, which currently is a hack for one special case of customizing the transformers. We could get rid of that argument, and instead you would specify `transformers=NormalizationTransformerGenerator(move_mean=False)`.

I like the `TransformerGenerator` idea. This issue with transformers requiring `Datasets` has come up a few times before and it feels like this is a useful abstraction. We could perhaps use the `TransformerGenerator` to provide naming information on transformers for caching. Perhaps to avoid filenames getting too long, we could have some caching directory structure like

```
featurizer_name/splitter_name/transformer_name
```

**Comment by peastman** [comment]
New version that allows Transformers to be specified.  As a bonus, the amount of dataset-specific code is now even less.

File: deepchem/molnet/load_function/delaney_datasets.py [hunk 7]
```diff
@@ -34,95 +51,34 @@ def load_delaney(featurizer='ECFP',
...
-  for transformer in transformers:
-    train = transformer.transform(train)
-    valid = transformer.transform(valid)
-    test = transformer.transform(test)
-
-  if reload:
-    deepchem.utils.data_utils.save_dataset_to_disk(save_folder, train, valid,
-                                                   test, transformers)
-  return delaney_tasks, (train, valid, test), transformers
+  loader = _DelaneyLoader(featurizer, splitter, transformers, data_dir,
```
Comment (rbharath) [inline deepchem/molnet/load_function/delaney_datasets.py:82]: Wow that's really slick!

File: deepchem/molnet/load_function/molnet_loader.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    self.transformer_class = transformer_class
+    self.kwargs = kwargs
+
+  def create_transformer(self, dataset: Dataset) -> dc.trans.Transformer:
+    """Construct a Transformer for a Dataset."""
+    return self.transformer_class(dataset=dataset, **self.kwargs)
+
+  def get_directory_name(self) -> str:
+    """Get a name for directories on disk describing this Transformer."""
+    name = self.transformer_class.__name__
```
Comment (rbharath) [inline deepchem/molnet/load_function/molnet_loader.py:41]: Should we add a string method for transformers the same way we have for splitters/featurizers? That might allow for more fine grained naming control in case there are transformer options that lead to different transformations

**Review by rbharath — COMMENTED** [review]
This is starting to look really slick! Have one minor comment about transformer string representations

File: deepchem/molnet/load_function/molnet_loader.py (hunk 8 above)
Comment (peastman) [reply deepchem/molnet/load_function/molnet_loader.py:41]: This is essentially the same string representation we use for splitters and featurizers: `classname_arg_value_arg_value`.  The code is a little different because we're describing a transformer that hasn't been created yet, so it can't go in the transformer's `__str__()` method.

**Comment by peastman** [comment]
Can anyone else reproduce the test failures from travis?  For example, it lists test_graph_models.py:test_graph_conv_regression_model as failing, but it works perfectly for me.

**Comment by peastman** [comment]
Ok, I figured it out.  All tests are now passing.  Shall we merge it?  We can then split up the work of converting all the other loader functions and do those in multiple PRs.

**Comment by rbharath** [comment]
@peastman Yes looks good to merge! Go for it once ready :)
